sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3574,15 +3568,64 @@ class Analyzer(override val catalogManager: CatalogManager)
 
   /**
    * Rule to mostly resolve, normalize and rewrite column names based on case sensitivity
-   * for alter table commands.
+   * for alter table column commands.
    */
-  object ResolveAlterTableCommands extends Rule[LogicalPlan] {
+  object ResolveAlterTableColumnCommands extends Rule[LogicalPlan] {
     def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
-      case a: AlterTableCommand if a.table.resolved && hasUnresolvedFieldName(a) =>
+      case a: AlterTableColumnCommand if a.table.resolved && hasUnresolvedFieldName(a) =>
         val table = a.table.asInstanceOf[ResolvedTable]
         a.transformExpressions {
-          case u: UnresolvedFieldName => resolveFieldNames(table, u.name, u)
+          case u: UnresolvedFieldName => resolveFieldNames(table, u.name, u.origin)
+        }
+
+      case a @ AlterTableAddColumns(r: ResolvedTable, cols) if hasUnresolvedColumns(cols) =>
```

Comment:
Note that since column name is `Seq[String]` this can be resolved many times if the column's position is None:
```
private def hasUnresolvedColumns(cols: Seq[QualifiedColType]): Boolean = {
  cols.exists(_.position.forall(!_.resolved))
}
```

Reply:
maybe we should do
```
case class QualifiedColType {
  path: Option[FieldName], // None for top-level columns
  colName: String,
  ...
}
```

Reply:
This looks fine. We can remove the `if hasUnresolvedColumns(cols)` which is not very useful here.

Reply:
Or, we can change `QualifiedColType`
```
case class QualifiedColType(
  path: FieldName,
  colName: String,
  ...
)
```

then we can keep the `Origin` of the column name, and save the resolution of the path at least.

Reply:
for top-level columns, `path` can be `ResolvedFieldName(Nil, StructField("root", tableSchema))`

Reply:
ping @imback82

Reply:
Sorry for the delay. I'll get to this today.

Reply:
OK, I will revert back to using `path: FieldName`.

Reply:
Updated `QualifiedColType`.`path` to be `FiledName` as suggested.